apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,8 +1395,221 @@ private void StartPush()
+                        if (extendIdsToRetry.Any())
+                        {
+                            // Retry for 10 seconds or 3 attempts only.
+                            RetryTemporaryFailures(Enumerable.Empty<string>(), null, extendIdsToRetry, extendsToRetry => _extendQueue.Locked(() => _extendQueue.Requeue(extendsToRetry)), 10);
+                        }
+                    }
+                    else
+                    {
+                        // We have acks or nacks.
+                        var idsToRetry = ackError.TemporaryFailureIds;
+                        var queue = ack.HasValue && ack.Value ? _ackQueue : _nackQueue;
+                        if (idsToRetry.Any())
+                        {
+                            // Retry for 600 seconds.
+                            RetryTemporaryFailures(idsToRetry, acksToRetry => queue.Locked(() => queue.Requeue(acksToRetry)), Enumerable.Empty<TimedId>(), null, 600);
+                        }
+                    }
+
+                    if (ackError.TryGetAcknowledgementException(out var exception))
+                    {
+                        if (hasExtendIds)
+                        {
+                            // Don't bubble permanent failures for lease extends as they are not user initiated.
+                            // Just log them.
+                            // TODO: Log the permanent failures.
+                            return;
+                        }
+
+                        // For acks/nacks, bubble up the exception.
+                        throw exception.FlattenIfPossible();
+                    }
+                }
+
+                // Retries the ackIds/nackIds with temporary failures for Acknowledge RPC or extendIds with temporary failures for ModifyAckDeadline RPC.
+                // This method checks if the id exists in _retryableIds dictionary and has not been tried for more than the specified timeout.
+                // If so, it retries those ids by adding them to the appropriate queue and calls the StartPush() method.
+                // For acks and nacks, ackOrnackIds must be non-null. For extends, extendIds must be non-null.
+                // For acks and nacks, ackActionToRetry must be non-null and specify the right queue (_ackQueue or _nackQueue).
+                // For extends, extendActionToRetry must be non-null and specify the right queue,i.e, _extendQueue.
+                void RetryTemporaryFailures(
+                    IEnumerable<string> ackOrNackIds,
+                    Action<IEnumerable<string>> ackActionToRetry,
+                    IEnumerable<TimedId> extendIds,
+                    Action<IEnumerable<TimedId>> extendActionToRetry = default,
+                    int timeoutInSeconds = 600)
+                {
+                    // TemporaryFailures may be empty but will never be null, so null check is not needed.
```

Comment:
TemporaryFailures => Temporary failures here and elsewhere - I suspect there was a find/replace that went wrong somewhere.